Firestore/core/src/firebase/firestore/local/local_store.mm

```diff
@@ -0,0 +1,486 @@
+}
+
+model::MaybeDocumentMap LocalStore::ApplyRemoteEvent(
+    const remote::RemoteEvent& remote_event) {
+  const SnapshotVersion& last_remote_version =
+      query_cache_->GetLastRemoteSnapshotVersion();
+
+  return persistence_->Run("Apply remote event", [&] {
+    // TODO(gsoltis): move the sequence number into the reference delegate.
+    ListenSequenceNumber sequenceNumber =
+        persistence_->current_sequence_number();
+
+    for (const auto& entry : remote_event.target_changes()) {
+      TargetId target_id = entry.first;
+      const TargetChange& change = entry.second;
+
+      auto found = target_ids.find(target_id);
+      if (found == target_ids.end()) {
+        // We don't update the remote keys if the query is not active. This
+        // ensures that we persist the updated query data along with the updated
+        // assignment.
+        continue;
+      }
+
+      QueryData old_query_data = found->second;
+
+      query_cache_->RemoveMatchingKeys(change.removed_documents(), target_id);
+      query_cache_->AddMatchingKeys(change.added_documents(), target_id);
+
+      // Update the resume token if the change includes one. Don't clear any
+      // preexisting value. Bump the sequence number as well, so that documents
+      // being removed now are ordered later than documents that were reviously
+      // _removed from this target.
+      const ByteString& resume_token = change.resume_token();
+      // Update the resume token if the change includes one.
+      if (!resume_token.empty()) {
+        QueryData new_query_data = old_query_data.Copy(
+            remote_event.snapshot_version(), resume_token, sequenceNumber);
+        target_ids[target_id] = new_query_data;
+
+        // Update the query data if there are target changes (or if sufficient
+        // time has passed since the last update).
+        if (ShouldPersistQueryData(new_query_data, old_query_data, change)) {
+          query_cache_->UpdateTarget(new_query_data);
+        }
+      }
+    }
+
+    OptionalMaybeDocumentMap changed_docs;
+    const DocumentKeySet& limbo_documents =
+        remote_event.limbo_document_changes();
+    DocumentKeySet updated_keys;
+    for (const auto& kv : remote_event.document_updates()) {
+      updated_keys = updated_keys.insert(kv.first);
+    }
+    // Each loop iteration only affects its "own" doc, so it's safe to get all
+    // the remote documents in advance in a single call.
+    OptionalMaybeDocumentMap existing_docs =
+        remote_document_cache_->GetAll(updated_keys);
+
+    for (const auto& kv : remote_event.document_updates()) {
+      const DocumentKey& key = kv.first;
+      const MaybeDocument& doc = kv.second;
+      absl::optional<MaybeDocument> existing_doc;
+      auto found_existing = existing_docs.get(key);
+      if (found_existing) {
+        existing_doc = *found_existing;
+      }
+
+      // Note: The order of the steps below is important, since we want to
```

Comment:
Hmm, this comment doesn't seem to exist in the version of `FSTLocalStore` that's in this branch. I presume it's the correct comment, but pointing it out to double-check whether everything is merged correctly.

Reply:
This is from https://github.com/firebase/firebase-ios-sdk/pull/3847, which is the fix to an issue related to index-free Sebastian found and fixed on Friday.